lib/load-config.js

```diff
@@ -22,30 +22,131 @@ const CONFIG_FILE_NAMES = Object.freeze([
   'nyc.config.cjs'
 ])
 
-function loadConfigFile (path) {
-  let config = {}
-  const jsExts = ['.js', '.cjs']
-  const ymlExts = ['.yml', '.yaml']
-  const fileName = basename(path)
+const EXTS_JS_VALIDATOR = /\.c?js$/i
+const EXTS_YAML_VALIDATOR = /\.ya?ml$/i
 
-  const ext = extname(path).toLowerCase()
-  if (jsExts.includes(ext)) {
-    config = require(path)
-  } else if (ymlExts.includes(ext)) {
-    config = require('js-yaml').load(readFileSync(path, 'utf8'))
-  } else if (ext === '.json' || fileName.slice(-2) === 'rc') {
-    config = JSON.parse(readFileSync(path, 'utf8'))
+class UnsupportedFileTypeError extends Error {
+  constructor (path) {
+    super(
+      `Unsupported file type ${extname(
+        path
+      ).toLowerCase()} while reading file ${path}`
+    )
+    this.name = UnsupportedFileTypeError.name
+  }
+}
+
+class ConfigParsingError extends Error {
+  constructor (path, errorDetails, originalError) {
+    super(
+      `Error loading configuration from file ${path}: ${errorDetails}` +
+        (typeof originalError === 'object' && originalError instanceof Error
+          ? ` Original error: ${originalError.message}`
+          : '')
+    )
+
+    this.name = UnsupportedFileTypeError.name
+
+    if (originalError instanceof Error) {
+      this.stack = originalError.stack
+    }
   }
+}
+
+const NO_EXPORTS = Symbol('no exports')
+
+function loadConfigFile (path, _di) {
+  const di = {
+    readFile: (path) => readFileSync(path, 'utf8'),
+    readJs: (path) => require(path),
+    ..._di
+  }
+
+  let config
+
+  const fileName = basename(path)
+  const ext = extname(path).toLowerCase()
+
+  if (EXTS_YAML_VALIDATOR.test(ext)) {
+    const JsYaml = require('js-yaml')
```

Comment:
Moved to the top of the file.  Just a preference.  

[Addressed](https://gist.github.com/mcknasty/80c8c814864241de05987dc3cae5fd54#file-parse-args-patch-diff-L126) in review branch.

Reply:
Moving to top of file is not allowed: `require('js-yaml')` MUST be executed late. This is because `js-yaml` is supposed to be a peer dependency and thus is only installed IF the user of c8 is using a YAML config file. Even then IIRC the original author of c8 wanted it loaded late, and that's why you'd originally placed this require down here.

Reply:
However I see that's it's a straight up dependency, and my memory of bcoe's requirements was wrong: https://github.com/bcoe/c8/pull/436#discussion_r1163150802. As such, requiring the library when this file is required is viable, but may create a minor amount of extra memory load vs late loading. Probably negligible and ignorable. I'm now OK with this change.